### PR TITLE
feat(gatsby): enable async rendering with react 18

### PR DESCRIPTION
**Pull request by wardpeet**
## Description

Enable Async rendering like Suspense during SSR. `pipeToNodeWritable` is a stream rendering that support Suspense boundaries and works with the new createRoot/hydrateRoot APIs.

https://github.com/reactwg/react-18/discussions/22

After ssr:
image

During client side navigation:
image

### CODE REVIEW

File: packages/gatsby/src/commands/build-html.ts [hunk 1]
```diff
@@ -234,7 +234,7 @@ const renderHTMLQueue = async (
           payload: pageSegment,
         })
 
-        for (const [_pagePath, arrayOfUsages] of Object.entries(
+        for (const [, arrayOfUsages] of Object.entries(
```
Comment (wardpeet) [inline packages/gatsby/src/commands/build-html.ts:237]: fix eslint warning

File: packages/gatsby/cache-dir/static-entry.js [hunk 2]
```diff
@@ -260,7 +264,27 @@ export default async function staticPage({
...
+          } = require(`./server-utils/writable-as-promise`)
+          const writableStream = new WritableAsPromise()
+          const { startWriting } = pipeToNodeWritable(
+            bodyComponent,
+            writableStream,
+            {
+              onCompleteAll() {
+                startWriting()
+              },
+              onError() {},
```
Comment (vladar) [inline packages/gatsby/cache-dir/static-entry.js:280]: Do we want to ignore errors here?

**Review by vladar — COMMENTED** [review]
Just to clarify - this doesn't use streaming per se for SSR, right? We aggregate changes and resolve full HTML as before. So is it just about working around the deprecation of `renderToString`?

File: packages/gatsby/cache-dir/static-entry.js (hunk 2 above)
Comment (wardpeet) [reply packages/gatsby/cache-dir/static-entry.js:280]: we won't do streaming yet until libs have adopted and maybe we never do.
React still throws the error so this is just to don't show the warning.

**Review by vladar — APPROVED** [review]

**Review by vladar — APPROVED** [review]

**Pull request merged**